docs/cardano-monads.rst

```diff
@@ -0,0 +1,608 @@
+async m = liftMockable $ Async m
+
+{-# INLINE withAsync #-}
+withAsync :: ( Mockable Async m ) => m t -> (Promise m t -> m r) -> m r
+withAsync mterm k = liftMockable $ WithAsync mterm k
+
+{-# INLINE wait #-}
+wait :: ( Mockable Async m ) => Promise m t -> m t
+wait promise = liftMockable $ Wait promise
+
+...
+```
+
+This increase is linear in the amount of supported methods.
+
+On the other hand, there's a decrease in the amount of boilerplate (linear in
+the amount of transformers), as each transformer needs to define
+``liftMockable`` just once.
+
+
+Method Dictionaries
+~~~~~~~~~~~~~~~~~~~
+
+``SendActions`` is another effect which is implemented differently from everything
+else in our system. It's an explicitly passed method dictionary that has the
+monad ``m`` as a type parameter. Unlike classes, explicit dictionaries must be passed
+manually (not by compiler via instance search), which might be both an advantage
+and a disadvantage.
+
+``SendActions`` is a perfect example of a *local* effect: in general, we don't
+have the ability to send requests, but we get this ability when we start a so
+called conversation. Doing a local effect via an explicit dictionary is a
+testament to poor support for local effects of our current approach with modes.
+
+One of the problems with ``SendActions`` is its use of natural transformations. We
+have a helper ``hoistSendActions``, and right now we horribly misuse it. Since
+``m`` in ``SendActions m`` is in an invariant position, we need
+``hoistSendActions`` to convert ``SendActions n`` to ``SendActions m``, and it
+requires two natural transformations: ``n ~> m`` and ``m ~> n``.
+
+First, providing these natural transformations is an inconvenience (the ``n ~> m`` one
+tends to be a mere lift, but the second one must do certain tricks to unlift).
+Second, it introduces a potential for horrible, subtle bugs. As was said, right
+now we misuse this helper, in particular because we call it in a runner,
+``runRealMode``; therefore, the unlift natural transformation must reconstruct the
+monadic context from what it has at transformer stack initialization site, not
+at conversation fork site; therefore, any local modifications to the monad
+transformer stack are *not* reflected in forked conversations (such as ``ReaderT``
+local).
+
+To put it into concrete terms, here is a code demonstration::
+
+    x <- ask
+    local f $ enqueueMsg msg $ \_ _ -> (:|[]) $ do
+      y <- ask
+
+What do you think, is ``y`` equal to ``x`` or ``f x``? The way we currently use
+``hoistSendActions``, it will be equal to ``x`` (completely oblivious to ``f``). This
+might be not at all what a programmer would expect.
+
+The moral of this story is that, perhaps, explicit dictionaries are a bad design
+because it's very easy to misuse them. A good effect system should take care of
+things like this. (But perhaps it's an over-genralization and it's only bad to unlift,
+whereas lifting is straightforward).
```

Comment:
I don't believe `SendActions m` should qualify as an effect. It wouldn't be right to say that a monad `m` determines a `SendActions `m, which is what *would* be said by having them determined by a type class instance on `m`. Each `SendActions m` value implicitly determines a destination (remote peer) as well as send and receive types, so if it were to be determined by its monad parameter then that type would also have to determine these things as well.

Reply:
> I don't believe SendActions m should qualify as an effect.

There's a definition of an effect in the beginning of the article, and `SendActions` fits this definition:

> Monadic effects extend the `Monad` interface with additional operations (such as `ask`, `local`, `get`, `put`, `throwError`, `liftIO`, `tell`, etc).

Indeed, `SendActions` extends the `Monad` interface with `withConnectionTo` and `enqueueMsg`.

> It wouldn't be right to say that a monad m determines a SendActionsm, which is what would be said by having them determined by a type class instance on m.

I am certain it would be possible to implement it this way. Basically, create a `ReaderT`-isomorphic transformer that provides `SendActions`. Then instead of passing `SendActions` manually one could put this transformer on top of the stack. `withConnectionTo` and `enqueueMsg` could be made into class methods and accessed from the dictionary provided by the transformer. To be clear, I am *not* advocating making this change, I only argue that `SendActions` is morally a monadic effect (and could be encoded conventionally).

Reply:
> I only argue that SendActions is morally a monadic effect (and could be encoded conventionally).

It's a value which constructs terms of some monad, i.e. constructs effects from existing effects. Would you make a new type class `MonadModify` for `modify :: Monad m => (s -> s) -> StateT s m ()` when it can be expressed in terms of `get` and `put`?

Reply:
It's not the same. You can't express SendActions only using operations of the base monad, otherwise its methods could be defined on top level. The reason is that there's peer info in its closure. Passing this info is an effect

Reply:
The only *effects* I need to express `SendActions m` are already in `m`. `SendActions` does not add any new irreducible effects. The values in the closure are used only to configure the existing effects, not to conjure up new ones.

Reply:
Suppose we delivered `SendActions m` via a typeclass

```Haskell
class HasSendActions m where
  sendActions :: m (SendActions m)
```

This is backwards. It says that the monad `m` determines a `SendActions m`, when in reality the particular `SendActions m` determines the monad which carries it. All we're doing here is replacing a `(->) (SendActions m)` with a `ReaderT (SendActions m)` for the perceived convenience of not having to pass a value, but at the cost of defining a new type class and instances wherever needed.

Reply:
> Passing this info is an effect

Surely passing information is not an effect. If it were, we'd have to say that `(+) 2 2` is an effectful term and so the definition of "pure" goes out the window.

Reply:
> Surely passing information is not an effect. 

Surely it is, even has a name — the reader effect.

>  If it were, we'd have to say that (+) 2 2 is an effectful term and so the definition of "pure" goes out the window.

"Pure" does not mean "without effects", it means "without SIDE effects". Consider `reader (uncurry (+)) :: Reader (Int, Int) Int`, it's an effectful value.

Reply:
> Surely it is, even has a name — the reader effect.
>
> "Pure" does not mean "without effects", it means "without SIDE effects". Consider reader (uncurry (+)) :: Reader (Int, Int) Int, it's an effectful value.

Is it an effectful term, given that it's exactly the same as the "uneffectful" term `uncurry (+) :: (Int, Int) -> Int`?

Referring to a reader as an effect is not useful and in fact only causes confusion. 

Where do you draw the line? When exactly is it OK to use typical function application, and when exactly should a `ReaderT` be thrown into the mix? Presumably there's good reason why we have `(+) :: Int -> Int -> Int` rather than `(+) :: Monad m => ReaderT Int (ReaderT Int) m Int` wouldn't you agree?

Reply:
> When exactly is it OK to use typical function application, and when exactly should a `ReaderT` be thrown into the mix?

I'd say that the difference is the depth of the call stack.

The `(+)` operation uses its arguments immediately, and when you call it it's unlikely that these arguments are from some context, so passing them explicitly is more convenient.

On the other hand, if `foo` calls `bar` calls `quux` calls `patak`, and a value is only used in `patak` (while other functions merely pass it downstream), this is asking for an effect (`Reader`) to automate this passing (in case we're already using another effect, such as IO, it's a cheap addition, as the code is already in `do`-notation; in pure code I would probably not make this tradeoff).

The difference between an effect and a side effect is whether it's reflected in the types.
The difference between an effect and not-an-effect is whether it's reflected in the terms. (OK, by this definition manual dictionary passing is not an effect, but I still consider it effect-related because the dictionaries contain effectful values inside).

Reply:
> The difference between an effect and a side effect is whether it's reflected in the types.
The difference between an effect and not-an-effect is whether it's reflected in the terms. (

Perhaps it has something to do with the type `m ()` and whether or not there's any term `mu :: m ()` such that `mu >>= (mf :: () -> m t) /= mf ()`. That's to say, can one monadic term influence later terms without passing on any information through the bind?

For all monads we have

```
  mu >>= (mf :: () -> m t)
= mu >>= (\_ -> mf ())
= mu >>  (mf ())
```

But only for some do we have

```
  forall mu m x . mu >> mx = mx
```

`Identity` being one such, and `ReaderT r m` as well whenever `m` also satisfies this equality. `IO`, `StateT s m`, `WriterT w m`, and `ExceptT e m` do not satisfy this equality, for any `m`.

That's why I wouldn't say reader is an effect. It doesn't introduce any "under the hood" interactions between sequenced terms. They're all dependent upon the value in the reader, but independent of *one another*.

Reply:
This is an insightful characterization. But let's be clear on the problem we have on hand — how to handle monadic effects or semi-effects (Reader/Identity). These are design choices we need to make: whether to pass explicitly/implicitly, whether to pass monadic values or non-monadic, whether to encode implementations in monad transformer stacks or not, etc.

I think we can conclude that "effect" is a bad choice of word for these questions, perhaps we should be saying "things that mention the `m` parameter" (or find a better term).